content/technotes/2020-07-07-github-pat-credentials.md

```diff
@@ -0,0 +1,99 @@
+
+In this post we show how you can take this one step further, and use the credentials package to save your `GITHUB_PAT` in the git credential store. This way you can authenticate with the GitHub API using the same token that is used for HTTPS remotes in git and gert. This is convenient for users, and also provides package authors with a mechanism to prompt the user for credentials, without having to take responsibility for managing tokens.
+
+
+## What is a Personal Access Token
+
+GitHub allows you to generate [Personal Access Tokens](https://github.com/settings/tokens), which you can use instead of your password when authenticating over HTTPS, both for git remotes and the GitHub API. Major advantages are:
+
+ - If you have enabled 2FA, you need a PAT to authenticate programmatically.
+ - A PAT can easily be revoked or updated. You can even automate this via the Github API.
+ - You can generate many PATs with specific permissions, giving you fine grained security control.
+
+In conclusion, if you are a responsible GitHub user, you have enabled 2FA on your account, and you only ever enter your main password when authenticating on the GitHub website. Everywhere else you should be using a PAT, preferably one that only has the permissions it needs.
+
+## Storing and loading your GitHub PAT
+
+Most R packages that interact with the GitHub API expect that your PAT is stored in the environment variable `GITHUB_PAT` in the R process. But how does it get there? The credentials package provides a function that will set this environment variable:
+
+```r
+# Tries to set the GITHUB_PAT environment variable
+credentials::set_github_pat()
+## TRUE
+```
+
+This function calls out to the [git credential store](https://git-scm.com/docs/git-credential-store) to get a suitable token for the `github.com` domain. If no token is available yet, the git credential manager will then prompt the user to enter one. What this looks like depends on your operating system and [credential helper](https://git-scm.com/docs/gitcredentials) configuration.
+
+![token](https://i.imgur.com/cnJcRmw.png)
+
+
+The `set_github_pat` function returns `TRUE` when it succeeded in setting the `GITHUB_PAT` environment variable, and `FALSE` if not. Packages that call `set_github_pat()` to let the user authenticate, can check the return value to determine if authentication was successful.
+
+Once a working PAT has been stored in the git credential store, it can automatically be loaded in another R session by calling `set_github_pat()` again. The token is automatically validated, and if it still works, the `GITHUB_PAT` environment variable is set without the user having to do anything.
```

Comment:
I didn't realize credentials validates the token! I will have to look at how you do this. Usethis also does this, but I'd love to delegate to credentials.